### ISSUE TITLE
AcousticEchoCanceler mode

### ISSUE THREAD

**Issue by snavruzov** [begeekmyfriend/yasea#343]
Hi, thanks for amazing project\
I noticed it's implemented built-in AEC support, but how can I test it if it's working in fact?\
I was doing the following steps:

- Start the stream.
- Open the stream in VLC.

Yasea and VLC are run in the same mobile device.\
But I still hear echo.

**Comment by snavruzov** [comment]
I'm using Samsung SM-G570F.\
Android version: 6.0.1

**Comment by begeekmyfriend** [comment]
Please check if this [line](https://github.com/begeekmyfriend/yasea/blob/master/library/src/main/java/net/ossrs/yasea/SrsPublisher.java#L78) was passed. If it is OK, try to expand the record buffer in [AudioRecord constructor](https://github.com/begeekmyfriend/yasea/commit/f6f071a6b624decfa6169e1841627df78152724d) in `SrsEncoder.java` or switch `AudioFormat.CHANNEL_IN_STEREO` to `AudioFormat.CHANNEL_IN_MONO`

**Comment by snavruzov** [comment]
Yes it's set to **true**, I put a log so it's supported, i multiplied buffersize to 8:
```
mic = new AudioRecord(MediaRecorder.AudioSource.VOICE_COMMUNICATION, SrsEncoder.ASAMPLERATE,
                    AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT, getPcmBufferSize() * 8);
```

I also tried VOICE_COMMUNICATION mode.\
One thing is not supported for some reason is _AutomaticGainControl.isAvailable()_ returning **false**

Stream delay between publisher and VLC player is ~5sec.

**Comment by snavruzov** [comment]
I commented onPause event to be able to run VLC and roll-up Yasea in background.
```
@Override
    protected void onPause() {
        super.onPause();
        //mPublisher.pauseRecord();
    }
```

**Comment by begeekmyfriend** [comment]
Let us try `AudioFormat.ENCODING_PCM_8BIT`. By the way, VLC generates delay itself.

**Comment by snavruzov** [comment]
Apparently my Samsung device doesn't support AEC, even it is says the mode enabled.\
I tested on another device Xiaomi Redmi 3S, and echo cancellation was working well.\

It's really disappointing to have this issue on Samsung devices, since I noticed the same behaviors on other Samsung models.\

Going further I'm researching software AEC implementation using WebRTC APM libs.\
So far I found a project https://github.com/dengzikun/WebRTC-APM-for-Android which uses audio only WEBRTC realization.

Currently I'm trying to integrate AECM module to be able to process audio through this module. There are some issues with audio packet buffers.\
I would be glad to contribute yasea adding this feature. I will update you once I get some results.\
Thanks.

**Comment by anytimehh** [comment]
使用该AEC算法要注意两点：

1）延时要小，因为算法默认滤波器长度是分为12块，每块64点，按照8000采样率，也就是12*8ms=96ms的数据，而且超过这个长度是处理不了的。

2）延时抖动要小，因为算法是默认10块也计算一次参考数据的位置（即滤波器能量最大的那一块），所以如果抖动很大的话找参考数据时不准确的，这样回声就消除不掉了。

**Comment by begeekmyfriend** [comment]
Hi all, yasea is a open project in my spare time for general usage. I have to admit there is somewhere imperfect. After prototype building period any extra advanced feature to be added would affect other's usage. In fact it has taken me a lot of time to adjust parameter like GOP, bit rate, FPS and so on to cater for all tastes. But still there are people complain on issues due to their particular network environment. I suggest we had better adjust these parameters to cater our extra requirements.

@anytimehh #325 may be helpful to you.

@snavruzov thank you for introducing `WebRTC-APM-for-Android`.

**Comment by begeekmyfriend** [comment]
> I commented onPause event to be able to run VLC and roll-up Yasea in background.

Good idea.

**Comment by snavruzov** [comment]
As I understand, in my case a built-in echo cancellation method works as following:
Everything coming from the speaker to the mic should be cut before wrapping to some audio codec.\
But in case of software AEC using here it seems doesn't work in that way since there is no AudioTrack instance used in yasea to process incoming speaker data.
Is there any way to retrieve(read) audio data coming from the speaker which is used by another app(MP3 app playing a music in background for example)?

**Comment by begeekmyfriend** [comment]
@snavruzov On this [Chinese Q&A site](http://www.dewen.net.cn/q/17437/Android), one talks about Samsung audio recording, when in stereo mode, the left channel is recording background voices while the right channel is recording human voices. It seems a bug of Samsung phone devices audio configuration in hardware driver.

**Comment by snavruzov** [comment]
Thank you for information, sure thing I was trying to use CHANNEL_IN_MONO and VOICE_COMMUNICATION modes, it works for almost all devices but Samsung.
However, for some reason it works for phone calls with speaker ON mode :). But why it doesn't want to work with AudioRecord libs.
So far, I'm trying to integrate WebRTC AECM module, but faced to some problems working with AudioTrack libs.

**Comment by begeekmyfriend** [comment]
Well if it is the bug of wrong configuration of audio driver that disables AEC, I doubt that you can make it even using some third-party modules.

**Comment by LauJJ** [comment]
I also encountered that some mobile phones do not support the native AEC function, so we can only extract the NS and AEC modules of webrtc for use
